lib/msprime.c

```diff
@@ -2631,17 +2644,58 @@ msp_recombination_event(msp_t *self, label_id_t label, segment_t **lhs, segment_
     return ret;
 }
 
+static double
+msp_generate_gc_tract_length(msp_t *self)
+{
+    double tl, continuous_gc_tract_length;
```

Comment:
Sorry to be slow to get around to this and for dragging it out more @fbaumdicker, but this looks really weird and unnatural to me. I've spent some time playing around with the values and how this is done, and I'm really confused now.

I thought that the *floor* of an exponential RV is geometrically distributed (note we're taking the ceiling here), and so we could generate the tract len value as an exponential with mean ``tract_length`` and then take the floor to get the required geometric with mean ``tract_length``. 

But this doesn't seem to be the case... I'm confused.

If we can't document the difference between the discrete and continuous case in one sentence, then we should probably just stick to discrete-only for GC.

Reply:
Hi @jeromekelleher,
this is slightly more complicated. In fact, the ceiling and the floor of an exponential RV are both geometrically distributed.
There are two types of geometric distributions. A) Number of trials until the first success appears and B) Number of trials before the first success. ceiling and floor result in A) and B), respectively.
For gene conversion tract length we want to use A) as each gene conversion tract should have at least length 1, and this is the distribution Wiuf and Hein used.
So this is why I am using the ceiling instead of the floor.

Furthermore, the ceiling of an exponential distributed RV with mean r is geometrically distributed but the mean is no longer given by r. See point 50 at https://www.randomservices.org/random/dist/Transformations.html .

In order to get the mean gc_tract_length for the ceiling of the exponential distributed rv we have to set the mean of the exponential distributed rv to
1 / (log(self->gc_tract_length) - log(self->gc_tract_length - 1))
(A similar formular holds for the floor)

The special limit case where gc_tract = 1, would lead to an exponential distributed RV with infinite rate, i.e. mean close to zero. In this case, the continuous gene conversion tract length is set to a very small value.

I have a branch with an updated gc documentation ready here https://github.com/fbaumdicker/msprime/commit/ca21a562a3c99b0f30dd90571dee217e220255f4. I was planning to create a documentation pull request after this one is merged.
However, currently, the docs are not explaining it as detailed as shown here. 
I will update the documentation to explain what is going on as detailed as here.
This should explain it properly. However, here is a one-sentence version:

_As a geometric distributed rv with mean r is equally distributed as the ceiling of an exponentially distributed rv with mean 1 / (log(r) - log(r - 1)), we are using exponentially distributed gene conversion tract lengths with the transformed mean when a continuous genome is simulated._


A related issue:
Are users often simulating genomes with a scaled length? I.e. using a genome length of 1 instead of N and scaling the recombination rate accordingly? Or is this a rare scenario?
I am asking as we currently do not support mean tract lengths below 1, which means the length of the continuous genome can currently not be scaled (together with the tract length) such that the tract length would be below 1.
However, I think it might be possible to allow this in the continuous case if we make a few more case distinctions, but personally I think it should suffice as it currently is, as you can easily simulate a continuous genome with length N without any disadvantages.

Reply:
Thanks very much for the explanation @fbaumdicker, this is super helpful. Since the distinction between the cases for discrete and continuous is rather subtle, and we're already making a if/else in any case, why don't we do somehting like
```
if (self->discrete_genome) {
    tl = gsl_ran_geometric(self->rng, 1 / self->tract_len);
} else 
    tl = gsl_ran_exponetial(self->rng, 1 / self->tract_len); // or whatever the param is that gives us mean = tract_len
}
```
I.e., in the discrete case we generate a geometric RV and in the continuous we generate an exponential?

Reply:
That is an excellent idea. This will make the code much more readable and should not make any difference. Will make the changes and push it in the next days.

Reply:
Fantastic, thanks @fbaumdicker! Would you mind also making the changes to the Python algorithms.py version - we want to keep the implementations reasonably in sync.

Reply:
Done!
Simplified the tract length generation and made the same changes to the python version.
Not sure about the lint error, I get now? What does it mean @jeromekelleher? Can we merge anyway?